File: leetcode/typescript/l00001_two_sum.ts
/**
 * 1. Two Sum
 * Easy - https://leetcode.com/problems/two-sum/
 * * Finds two numbers in an array that sum to a specific target using a one-pass Map.
 * * @param nums An array of integers.
 * @param target The integer sum to search for.
 * @returns An array containing the two indices, or an empty array if no solution exists.
 */
export function solve(nums: number[], target: number): number[] {
    // Maps value -> index
    const seen = new Map<number, number>();

    for (let i = 0; i < nums.length; i++) {
        const diff = target - nums[i];

        if (seen.has(diff)) {
            // map.get() is guaranteed to return a number here since seen.has() passed
            return [seen.get(diff)!, i];
        }
        seen.set(nums[i], i);
    }

    return [];
}

File: leetcode/rust/src/l00001_two_sum.rs
use std::collections::HashMap;

pub struct Solution;

impl Solution {
    /// Finds two numbers in a vector that sum to a specific target using a one-pass hash map.
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        // Maps value -> index
        let mut seen = HashMap::new();

        for (i, &num) in nums.iter().enumerate() {
            let diff = target - num;

            if let Some(&index) = seen.get(&diff) {
                return vec![index as i32, i as i32];
            }
            seen.insert(num, i);
        }

        // Return empty vector if no match is found
        vec![]
    }
}